tests/agent_disconnect_test.go

```diff
@@ -0,0 +1,151 @@
+package tests
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+	"sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client"
+)
+
+func TestProxy_Agent_Disconnect_Same_Stream(t *testing.T) {
+	server := httptest.NewServer(newEchoServer("hello"))
+	defer server.Close()
+
+	stopCh := make(chan struct{})
+
+	proxy, cleanup, err := runGRPCProxyServer()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
+
+	runAgent(proxy.agent, stopCh)
+
+	// Wait for agent to register on proxy server
+	time.Sleep(time.Second)
+
+	// run test client
+	tunnel, err := client.CreateSingleUseGrpcTunnel(proxy.front, grpc.WithInsecure())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c := &http.Client{
+		Transport: &http.Transport{
+			Dial: tunnel.Dial,
+		},
+	}
+
+	verify := func(expectSuccess bool) {
+
+		r, err := c.Get(server.URL)
+
+		if expectSuccess {
+			if err != nil {
+				t.Error(err)
+			}
+
+			data, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				t.Error(err)
+			}
+			defer r.Body.Close()
+
+			if string(data) != "hello" {
+				t.Errorf("expected test, got %v", data)
+			}
+		} else {
+			if err == nil {
+				t.Error("expect second request on broken agent connection to fail")
+			}
+		}
+
+	}
+
+	verify(true)
+	close(stopCh)
+
+	// It takes up to 5 seconds for the agent to recognize the channel has been closed and shut down all connections.
+	time.Sleep(7 * time.Second)
+	verify(false)
+}
+
+func TestProxy_Agent_Reconnect(t *testing.T) {
+	server := httptest.NewServer(newEchoServer("hello"))
+	defer server.Close()
+
+	stopCh := make(chan struct{})
+
+	proxy, cleanup, err := runGRPCProxyServer()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
+
+	runAgent(proxy.agent, stopCh)
+
+	// Wait for agent to register on proxy server
+	time.Sleep(time.Second)
+
+	// run test client
+
+	verify := func(expectSuccess bool) {
+
+		tunnel, err := client.CreateSingleUseGrpcTunnel(proxy.front, grpc.WithInsecure())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		c := &http.Client{
+			Transport: &http.Transport{
+				Dial: tunnel.Dial,
+			},
+		}
+
+		r, err := c.Get(server.URL)
+
+		if expectSuccess {
+			if err != nil {
+				t.Error(err)
+			}
+
+			data, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				t.Error(err)
+			}
+			defer r.Body.Close()
+
+			if string(data) != "hello" {
+				t.Errorf("expected test, got %v", data)
+			}
+		} else {
+			if err == nil {
+				t.Error("expect second request on broken agent connection to fail")
+			}
+		}
+
+	}
+
+	verify(true)
+	close(stopCh)
+	// It takes up to 5 seconds for the agent to recognize the channel has been closed and shut down all connections.
+	time.Sleep(7 * time.Second)
```

Comment:
This looks fragile. How about modify this to a waitUntil?

Also why would it take up to 5 seconds? I expect it to be <1s.

Reply:
Thanks for the catch, yeah it should be less than 1s. I tested this multiple times with 1s and it passes without any problems. I think it should be fine given that we also do `time.Sleep(time.Second)` in other places of our test code.

Reply:
If verify is idempotent we could consider using https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L420 or waitUntil. It will allow us to test quickly but be resilient to problems with the test running slowly. We already have a dependency upon apimachinery, so I don't expect that to be a problem.

Reply:
verify is idempotent in the sense that the output is the same, but has side effects (eg: golang caching http 1.1 persistent connections) that I'd like to avoid. If we want to poll, I think it should be done on the proxy server to determine the status of its ongoing connections. Currently this kind of function doesn't exist, but I think it would be useful to have for debugging purposes.

Reply:
I'm mostly thinking of it as way to early out in a test when a condition is met. So we can check a condition every second until it passes or some time out (30 seconds). Most of the time the test will run very quickly but when it does slow down it is less likely to become a flaky test.

Reply:
I'm more concerned about the test becoming flaky. I hope one day we can get rid of the 1s sleep as well. If a waitUntil can work here, let's avoid introducing more latency sensitive code.

RE. "golang caching http 1.1 persistent connections", sorry I didn't get why you want to avoid that in this test?

Reply:
You can control the connection cache by creating or not creating new tunnel, right?

Reply:
It's not a concern in the current test, but might cause debugging to be harder down the line if it becomes a standard. Logs also get polluted which is undesirable.

I updated this to use an alternate method,  polling for for proxy server readiness instead. I think this is a cleaner way to express what we want, let me know if you have additional suggestions. We can also replace the 1s sleeps with this function as well.

Reply:
I removed the test for the failure when no backend(agent) is available since this PR isn't targeting dial failures. We handle dial failures by just waiting until the 30s dial timeout, and optimizing this is something that should be done in another PR. (Incorporating this test significantly slows down the test runtime at the moment)

Reply:
Yeah, polling the readiness is much better. That's the right way to replace the 1s sleeps. We can do that in a followup.